must-call-checker/README.md

```diff
@@ -0,0 +1,88 @@
+### MustCall Checker
+
+The MustCall Checker tracks the obligations that an object might have to call particular methods
+before it is deallocated. For example, an object whose type is `java.net.Socket` might
+have an obligation to call the `close()` method, to release the socket. Or, a builder object
+might have an obligation to call the `build()` method. We say that an object `obj` has a
+*must-call obligation* to call some method `foo` if there exists an execution path on which
+it is an error for `obj.foo()` not to be invoked before `obj` is deallocated. As a shorthand,
+the documentation for this checker sometimes shortens this to "`obj` must call `foo`" or similar,
+to refer to the worst case in which `obj` is actually an object that has to call `foo` (even if
+`obj` sometimes dynamically does not actually need to call `foo`, such as when `obj` is `null`).
+
+The MustCall Checker tracks which methods might need to be called, but does
+**not** check that these methods have actually been called.  The Object
+Construction Checker's `-AcheckMustCall` command-line option checks the
+obligation expressed by an `@MustCall` annotation. This option uses a dataflow
+algorithm to compare `@MustCall` obligations to inferred `@CalledMethods` facts,
+and reports cases where required methods might not have been called.
+
+
+### Explanation of qualifiers and type hierarchy
+
+The MustCall Checker supports two qualifiers: `@MustCall` and `@MustCallTop`. The `@MustCall`
+qualifier's arguments are the methods that the annotated type
+might be responsible for calling. The type `@MustCall({"a"}) Object obj` means
+that at run time `obj` cannot have an obligation to call any method other than `a()` (but it may or 
+may not actually have an obligation to call `a()`).
+Likewise, `@MustCall({"m1", "m2"}) Object` represents all objects that need to
+call `m1`, `m2`, both, or neither. It should not represent an object that needs
+to call some *other* method.  The type hierarchy is:
+
+                     @MustCallTop
+                           |
+                          ...
+                           |
+               @MustCall({"m1", "m2"})     ...
+                    /             \
+       @MustCall({"m1"})     @MustCall({"m2"})  ...
+                    \            /
+                    @MustCall({})
+
+Note that `@MustCall({"m1", "m2"}) Object` represents more possible values than
+`@MustCall({"m1"}) Object` does, because `@MustCall({"m1"}) Object` can only
+contain an object that needs to call `m1` or one that needs to call nothing - an
+object that needs to call `m2` (or both `m1` and `m2`) is not permitted.
+
+The default type in unannotated code is `@MustCall({})`.  Given an expression of
+unannotated type, its value definitely has no obligations to call any methods.
+
+This default is unsound in the sense that a priori the checker
```

Comment:
> A default cannot be unsound

I disagree with this statement. My disagreement, I think, hinges on what soundness means. 

One way to define soundness is that the type system never approves a program that is inconsistent with its annotations. I think that's the definition you are using.

Another way to define soundness is that there is no way to annotate an incorrect program with a set of consistent annotations. That's what I mean here. The problem is that for the must-call type system (and for some others we've talked about extensively, like data classification), there is no way to know a priori what the right defaults are, and taking the most conservative approach (`@MustCallAny` as the default for this type system, or `@Critical` for data classification, etc) isn't useful. Using `@MustCall({})` as the default violates this notion of soundness (but not the other notion) because we, as humans, know that some classes (like `Socket`) do actually have must-call obligations.

For many of the other type checkers we've built, this isn't a concern, because there is an objective truth that we can easily know ahead of time and build into the checker. For example, the Index Checker can know definitively what is and is not an array access, and therefore can guarantee both of these notions of soundness at the same time.

Reply:
> Another way to define soundness is that there is no way to annotate an incorrect program with a set of consistent annotations.

Can you define "incorrect program", as you mean it here?  Thanks.

Reply:
I've removed this paragraph in my latest draft, as you originally suggested, based on our call earlier.

However, I think this question is independently interesting, so I'm going to try to answer it anyway. By "incorrect program" here I mean a program that does something that the checker promised wouldn't be possible if the program typechecks. The distinction I'm trying to get at between a checker like this one and e.g. the nullness or index checker is that this checker makes a guarantee **modulo a user-provided specification**, whereas those checkers make a guarantee, full stop. If the nullness checker approves your program (while being able to see the full program source) without issuing an error, the program definitely will not dereference a null pointer. By contrast, even with access to the full program text, the must-call checker (with the Object Construction Checker's `-AcheckMustCall` option) cannot notice that a required method was not called unless a user tells the checker that that method needs to be called.